datafusion/physical-expr/src/expressions/case.rs

```diff
@@ -256,6 +289,36 @@ impl CaseExpr {
 
         Ok(ColumnarValue::Array(current_value))
     }
+
+    /// This function evaluates the specialized case of:
+    ///
+    /// CASE WHEN condition THEN column
+    ///      [ELSE NULL]
+    /// END
+    fn case_column_or_null(&self, batch: &RecordBatch) -> Result<ColumnarValue> {
+        let when_expr = &self.when_then_expr[0].0;
+        let then_expr = &self.when_then_expr[0].1;
+        if let ColumnarValue::Array(bit_mask) = when_expr.evaluate(batch)? {
```

Comment:
If it is evaluated to `Scalar`, maybe we still can work on it? I.e., we can convert the scalar value to an boolean array.

Reply:
I am planning on adding a specialization for scalar values as well to avoid converting scalars to arrays

Reply:
actually, nm, your question is different. Yes, we could implement special handling for `CASE WHEN [true|false] THEN` but I'm not sure that is a real-world use case that is worth optimizing

Reply:
Oh, I forgot that in this optimization, the when expression could only be a `Column`.